### ISSUE TITLE
Kubelet fails to retrieve secret with empty name

### ISSUE THREAD

**Issue by polycaster** [kubernetes/kubernetes#99454]
#### What happened:
Kubelet keeps trying to retrieve secrets for "non-empty namespace"/"empty secret name" and this is generating very large amounts of logs

#### What you expected to happen:
I would expect kubelet to not try to retrieve secrets with an empty name.

#### How to reproduce it (as minimally and precisely as possible):
Unsure, I'm unable to determine the source that is causing kubelet to output these log lines.

#### Anything else we need to know?:
In a cluster with 330 nodes i'm seeing roughly 3 million of these messages per half hour:
```
E0225 09:23:21.876843    1230 reflector.go:125] object-"seahorse"/"": Failed to list *v1.Secret: secrets is forbidden: User "system:node:$node_SA_identifier" cannot list resource "secrets" in API group "" in the namespace "seahorse": No Object name found
E0225 09:23:21.758400    1309 reflector.go:125] object-"arcadia"/"": Failed to list *v1.Secret: secrets is forbidden: User "system:node:$node_SA_identifier" cannot list resource "secrets" in API group "" in the namespace "arcadia": No Object name found
E0225 09:23:21.726618    1272 reflector.go:125] object-"inception"/"": Failed to list *v1.Secret: secrets is forbidden: User "system:node:$node_SA_identifier" cannot list resource "secrets" in API group "" in the namespace "inception": No Object name found
E0225 09:23:21.709821    1301 reflector.go:125] object-"istio-system"/"": Failed to list *v1.Secret: secrets is forbidden: User "system:node:$node_SA_identifier" cannot list resource "secrets" in API group "" in the namespace "istio-system": No Object name found
```
These messages show up for a lot of namespaces but not all.

**If the node attempts to retrieve a secret with a name it succeeds, which rules out RBAC issues. I believe the `secrets is forbidden` is a red-herring**
I've checked all secrets, in case there was somehow one with no name, and couldn't find anything.

#### Environment:
- Kubernetes version (use `kubectl version`):  
```
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.15-gke.7800", GitCommit:"cef3156c566a1d1a4b23ee360a760f45bfbaaac1", GitTreeState:"clean", BuildDate:"2020-12-14T09:12:37Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}

and 

Server Version: version.Info{Major:"1", Minor:"17+", GitVersion:"v1.17.12-eks-7684af", GitCommit:"7684af4ac41370dd109ac13817023cb8063e3d45", GitTreeState:"clean", BuildDate:"2020-10-20T22:57:40Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: AWS and GCP
- OS (e.g: `cat /etc/os-release`): ` Amazon Linux 2 and Container-Optimized OS from Google`
- Kernel (e.g. `uname -a`): `4.14.186-146.268.amzn2.x86_64 and 4.19.112+`
- Install tools: `N/A`
- Network plugin and version (if this is a network-related bug): `N/A`
- Others:
```
Kubelet Version:             v1.17.9-eks-4c6976
and
Kubelet Version:            v1.15.12-gke.20
```

**Comment by liggitt** [comment]
cc @wojtek-t - looks like the informer-based secret manager

**Comment by liggitt** [comment]
The forbidden message comes from here:

https://github.com/kubernetes/kubernetes/blob/bd190762fb77b78788e1a3ba72d70a8a08e184b8/plugin/pkg/auth/authorizer/node/node_authorizer.go#L196-L199

Nodes are not authorized to watch all secrets, only specifically named ones.

The watch is being attempted here:

https://github.com/kubernetes/kubernetes/blob/bd190762fb77b78788e1a3ba72d70a8a08e184b8/pkg/kubelet/util/manager/watch_based_manager.go#L118-L124

It is unexpected that code path would be called with an empty name, so figuring out what led to that would be the next step.

Is this reproducible on a currently supported release branch (1.18+)?

**Comment by liggitt** [comment]
It looks like API validation permits some of the `secretRef` pod fields to be non-nil without specifying a secret name, and the kubelet code that starts informers to watch those secrets was not correctly filterering those out. When the kubelet tried to list/watch secrets without specifying a name, it was forbidden (correctly), but produced those error messages. https://github.com/kubernetes/kubernetes/pull/99538 will filter out empty secret names and fix the log errors.

In the meantime, you can resolve the issue on your end by looking at the four namespaces in the logs, finding the workloads or pods specifying a non-nil `imagePullSecrets` / `secret` / `secretRef` / `secretKeyRef` field but are not providing a secret name, and removing the ineffective field.

**Comment by wojtek-t** [comment]
Thanks Jordan for debugging and fixing!

**Comment by polycaster** [comment]
Just an FYI for anyone that might see this issue on their environment.

I found that Istio was injecting sidecar proxies with empty `imagePullSecrets`. I was unable to see this by describing `Deployment`, `Pod`, and `ReplicaSet`.
The only way of seeing the relevant empty field was to edit the pod at which point I was able to see the following:

```
    name: istio-proxy
    ports:
    - containerPort: 15090
      name: http-envoy-prom
      protocol: TCP
    readinessProbe:
      failureThreshold: 30
      httpGet:
        path: /healthz/ready
        port: 15021
        scheme: HTTP
      initialDelaySeconds: 1
      periodSeconds: 2
      successThreshold: 1
      timeoutSeconds: 1
    resources:
      limits:
        cpu: "2"
        memory: 1Gi
      requests:
        cpu: 100m
        memory: 128Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: true
      runAsGroup: 1337
      runAsNonRoot: true
      runAsUser: 1337
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/istio
      name: istiod-ca-cert
    - mountPath: /var/lib/istio/data
      name: istio-data
    - mountPath: /etc/istio/proxy
      name: istio-envoy
    - mountPath: /etc/istio/pod
      name: istio-podinfo
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-fcgdr
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  imagePullSecrets:
  - {}
 ```

**Comment by liggitt** [comment]
> I found that Istio was injecting sidecar proxies with empty `imagePullSecrets`. I was unable to see this by describing `Deployment`, `Pod`, and `ReplicaSet`.

Thanks for digging. Has this been reported to Istio?

**Comment by polycaster** [comment]
If istio were deployed with the default args it wouldn't face this issue.

What I found on my side is that we started doing some templating around istio's arguments for `imagePullSecrets`. The default value of this flag is `[]` but the way it was being initialised when there were no secrets was `[{}]` which explains the lack of name for the secret.

I've confirmed since that if the default value is used these messages never surface, so all good.
